backend/provisioner/inmem_provisioner.go

```diff
@@ -0,0 +1,157 @@
+
+// InMemProvisioner for running an in memory provisioner, constructing all resources concurrently
+//
+// It spawns a separate goroutine for each resource to be provisioned, and
+// finishes the task when all resources are provisioned or an error occurs.
+type InMemProvisioner struct {
+	running  *xsync.MapOf[string, *inMemProvisioningTask]
+	handlers map[ResourceType]InMemResourceProvisionerFn
+}
+
+func NewEmbeddedProvisioner(handlers map[ResourceType]InMemResourceProvisionerFn) *InMemProvisioner {
+	return &InMemProvisioner{
+		running:  xsync.NewMapOf[string, *inMemProvisioningTask](),
+		handlers: handlers,
+	}
+}
+
+var _ provisionerconnect.ProvisionerPluginServiceClient = (*InMemProvisioner)(nil)
+
+func (d *InMemProvisioner) Ping(context.Context, *connect.Request[ftlv1.PingRequest]) (*connect.Response[ftlv1.PingResponse], error) {
+	return &connect.Response[ftlv1.PingResponse]{}, nil
+}
+
+func (d *InMemProvisioner) Plan(context.Context, *connect.Request[provisioner.PlanRequest]) (*connect.Response[provisioner.PlanResponse], error) {
+	panic("unimplemented")
+}
+
+func (d *InMemProvisioner) Provision(ctx context.Context, req *connect.Request[provisioner.ProvisionRequest]) (*connect.Response[provisioner.ProvisionResponse], error) {
+	logger := log.FromContext(ctx)
+
+	previous := map[string]*provisioner.Resource{}
+	for _, r := range req.Msg.ExistingResources {
+		previous[r.ResourceId] = r
+	}
+
+	task := &inMemProvisioningTask{}
+	for _, r := range req.Msg.DesiredResources {
+		if handler, ok := d.handlers[TypeOf(r.Resource)]; ok {
+			step := &inMemProvisioningStep{Resource: r.Resource}
+			task.steps = append(task.steps, step)
+			go func() {
+				defer step.Done.Store(true)
+				output, err := handler(ctx, r, req.Msg.Module, r.Resource.ResourceId)
+				if err != nil {
+					step.Err = err
+					logger.Errorf(err, "failed to provision resource %s", r.Resource.ResourceId)
+					return
+				}
+				step.Resource = output
+			}()
+		} else {
+			err := fmt.Errorf("unsupported resource type: %T", r.Resource.Resource)
+			return nil, connect.NewError(connect.CodeInvalidArgument, err)
+		}
+	}
+
+	if len(task.steps) == 0 {
+		return connect.NewResponse(&provisioner.ProvisionResponse{
+			Status: provisioner.ProvisionResponse_NO_CHANGES,
+		}), nil
+	}
+
+	token := uuid.New().String()
+	logger.Debugf("started a task with token %s", token)
+	d.running.Store(token, task)
+
+	return connect.NewResponse(&provisioner.ProvisionResponse{
+		ProvisioningToken: token,
+		Status:            provisioner.ProvisionResponse_SUBMITTED,
+	}), nil
+}
+
+func (d *InMemProvisioner) Status(ctx context.Context, req *connect.Request[provisioner.StatusRequest]) (*connect.Response[provisioner.StatusResponse], error) {
+	logger := log.FromContext(ctx)
+
+	token := req.Msg.ProvisioningToken
+	task, ok := d.running.Load(token)
+	if !ok {
+		return statusFailure(fmt.Sprintf("unknown token: %s", token))
```

Comment:
Should this return the `invalid_argument` connect error code instead of `unknown` as `statusFailure` does?

Reply:
Good point, will change!